jerry-ext/jerry-module/jerry-module.c

```diff
@@ -0,0 +1,192 @@
+
+  if (return_value)
+  {
+    if (previous)
+    {
+      previous->next = return_value->next;
+    }
+    else
+    {
+      (*root) = return_value->next;
+    }
+    JERRYX_MODULE_LINK_FLOAT (return_value);
+  }
+
+  return return_value;
+} /* jerryx_module_link_remove */
+
+static void
+jerryx_module_link_free (jerryx_module_header_t *root,
+                         void (*free_link) (jerryx_module_header_t *link))
+{
+  jerryx_module_header_t *next;
+  while (root)
+  {
+    next = root->next;
+    free_link (root);
+    root = next;
+  }
+} /* jerryx_module_link_free */
+
+typedef struct
+{
+  jerryx_module_header_t link;
+  jerry_value_t result;
+} jerryx_module_instance_t;
+
+static void
+jerryx_module_free_instance (jerryx_module_header_t *link)
+{
+  jerryx_module_instance_t *instance = (jerryx_module_instance_t *) link;
+  jerry_release_value (instance->result);
+  jmem_heap_free_block (instance, sizeof (jerryx_module_instance_t));
+} /* jerryx_module_free_instance */
+
+void *
+jerryx_module_context_init (void)
+{
+  void *return_value = jmem_heap_alloc_block_null_on_error (sizeof (jerryx_module_instance_t *));
+  memset (return_value, 0, sizeof (jerryx_module_instance_t *));
+  return return_value;
+} /* jerryx_module_context_init */
+
+void
+jerryx_module_context_deinit (void *context_p)
+{
+  if (context_p)
+  {
+    jerryx_module_link_free (*((jerryx_module_header_t **) (context_p)), jerryx_module_free_instance);
+    jmem_heap_free_block (context_p, sizeof (jerryx_module_header_t *));
+  }
+} /* jerryx_module_context_deinit */
+
+#ifdef JERRYX_MODULE_HAVE_CONTEXT
+#include "jerry-context.h"
+static int jerryx_module_user_context_slot = -1;
+JERRYX_C_CTOR (_register_module_handler_)
+{
+  jerryx_module_user_context_slot = jerryx_context_request_slot (jerryx_module_context_init,
+                                                                 jerryx_module_context_deinit);
+} /* JERRYX_C_CTOR */
+#define JERRYX_MODULE_CONTEXT \
+  (jerryx_context_get (jerryx_module_user_context_slot))
+#else /* !JERRYX_MODULE_HAVE_CONTEXT */
+#define JERRYX_MODULE_CONTEXT \
+  (jerry_get_user_context ())
+#endif /* JERRYX_MODULE_HAVE_CONTEXT */
+
+/* Context-free list of modules */
+static jerryx_module_t *modules = NULL;
+
+void
+jerryx_module_register (jerryx_module_t *module_p)
+{
+  jerryx_module_link_insert ((jerryx_module_header_t *) module_p, ((jerryx_module_header_t **) (&modules)));
+} /* jerryx_module_register */
+
+void
+jerryx_module_unregister (jerryx_module_t *module_p)
+{
+  jerryx_module_link_remove (module_p->link.name, ((jerryx_module_header_t **) (&modules)));
+} /* jerryx_module_unregister */
+
+typedef struct
+{
+  jerryx_module_header_t link;
+  jerry_value_t result;
+} jerryx_instance_t;
+
+static const char not_found_prologue[] = "Module '";
+static const char not_found_epilogue[] = "' not found";
+
+jerry_value_t
+jerryx_module_load (const jerry_char_t *name)
```

Comment:
It's not a problem. Quite the contrary: The requirement to have a cache (because otherwise you cannot have idempotency) supports your argument to `jerry_port_fatal ()` if we fail to allocate the memory for the module instance. That's why I switched to `jmem_heap_alloc_block ()`.